File: farebot-base/src/main/java/com/codebutler/farebot/base/util/ByteUtils.java
package com.codebutler.farebot.base.util;

import java.math.BigInteger;

public final class ByteUtils {

    private ByteUtils() { }

    public static String getHexString(byte[] b) {
        String result = "";
        for (int i = 0; i < b.length; i++) {
            result += Integer.toString((b[i] & 0xff) + 0x100, 16).substring(1);
        }
        return result;
    }

    public static String getHexString(byte[] b, String defaultResult) {
        try {
            return getHexString(b);
        } catch (Exception ex) {
            return defaultResult;
        }
    }

    public static byte[] hexStringToByteArray(String s) {
        if ((s.length() % 2) != 0) {
            throw new IllegalArgumentException("Bad input string: " + s);
        }

        int len = s.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                    + Character.digit(s.charAt(i + 1), 16));
        }
        return data;
    }

    public static byte[] intToByteArray(int value) {
        return new byte[] {
                (byte)(value >>> 24),
                (byte)(value >>> 16),
                (byte)(value >>> 8),
                (byte)value};
    }

    public static int byteArrayToInt(byte[] b) {
        return byteArrayToInt(b, 0);
    }

    private static int byteArrayToInt(byte[] b, int offset) {
        return byteArrayToInt(b, offset, b.length);
    }

    public static int byteArrayToInt(byte[] b, int offset, int length) {
        return (int) byteArrayToLong(b, offset, length);
    }

    public static long byteArrayToLong(byte[] b, int offset, int length) {
        if (b.length < offset + length) {
            throw new IllegalArgumentException("offset + length must be less than or equal to b.length");
        }

        long value = 0;
        for (int i = 0; i < length; i++) {
            int shift = (length - 1 - i) * 8;
            value += (b[i + offset] & 0x000000FF) << shift;
        }
        return value;
    }

    public static BigInteger byteArrayToBigInteger(byte[] b, int offset, int length) {
        if (b.length < offset + length) {
            throw new IllegalArgumentException("offset + length must be less than or equal to b.length");
        }

        BigInteger value = BigInteger.valueOf(0);
        for (int i = 0; i < length; i++) {
            value = value.shiftLeft(8);
            value = value.add(BigInteger.valueOf(b[i + offset] & 0x000000ff));
        }
        return value;
    }

    public static byte[] byteArraySlice(byte[] b, int offset, int length) {
        byte[] ret = new byte[length];
        System.arraycopy(b, offset, ret, 0, length);
        return ret;
    }

    public static int convertBCDtoInteger(byte data) {
        return (((data & (char) 0xF0) >> 4) * 10) + ((data & (char) 0x0F));
    }

    public static int getBitsFromInteger(int buffer, int iStartBit, int iLength) {
        return (buffer >> (iStartBit)) & ((char) 0xFF >> (8 - iLength));
    }

    /**
     * Reverses a byte array, such that the last byte is first, and the first byte is last.
     *
     * @param buffer     Source buffer to reverse
     * @param iStartByte Start position in the buffer to read from
     * @param iLength    Number of bytes to read
     * @return A new byte array, of length iLength, with the bytes reversed
     */
    public static byte[] reverseBuffer(byte[] buffer, int iStartByte, int iLength) {
        byte[] reversed = new byte[iLength];
        int iEndByte = iStartByte + iLength;
        for (int x = 0; x < iLength; x++) {
            reversed[x] = buffer[iEndByte - x - 1];
        }
        return reversed;
    }

    /**
     * Given an unsigned integer value, calculate the two's complement of the value if it is
     * actually a negative value
     *
     * @param input      Input value to convert
     * @param highestBit The position of the highest bit in the number, 0-indexed.
     * @return A signed integer containing it's converted value.
     */
    public static int unsignedToTwoComplement(int input, int highestBit) {
        if (getBitsFromInteger(input, highestBit, 1) == 1) {
            // inverse all bits
            input ^= (2 << highestBit) - 1;
            return -(1 + input);
        }

        return input;
    }

    /* Based on function from mfocGUI by 'Huuf' (http://www.huuf.info/OV/) */
    public static int getBitsFromBuffer(byte[] buffer, int iStartBit, int iLength) {
        // Note: Assumes big-endian
        int iEndBit = iStartBit + iLength - 1;
        int iSByte = iStartBit / 8;
        int iSBit = iStartBit % 8;
        int iEByte = iEndBit / 8;
        int iEBit = iEndBit % 8;

        if (iSByte == iEByte) {
            return ((char) buffer[iEByte] >> (7 - iEBit)) & ((char) 0xFF >> (8 - iLength));
        } else {
            int uRet = (((char) buffer[iSByte] & (char) ((char) 0xFF >> iSBit)) << (((iEByte - iSByte - 1) * 8)
                    + (iEBit + 1)));

            for (int i = iSByte + 1; i < iEByte; i++) {
                uRet |= (((char) buffer[i] & (char) 0xFF) << (((iEByte - i - 1) * 8) + (iEBit + 1)));
            }

            uRet |= (((char) buffer[iEByte] & (char) 0xFF)) >> (7 - iEBit);

            return uRet;
        }
    }
}

File: base/src/commonMain/kotlin/com/codebutler/farebot/base/util/ByteUtils.kt
package com.codebutler.farebot.base.util

object ByteUtils {
    fun getHexString(b: ByteArray): String {
        val sb = StringBuilder(b.size * 2)
        for (byte in b) {
            val v = byte.toInt() and 0xFF
            sb.append(HEX_CHARS[v ushr 4])
            sb.append(HEX_CHARS[v and 0x0F])
        }
        return sb.toString()
    }

    fun getHexString(
        b: ByteArray,
        @Suppress("UNUSED_PARAMETER") defaultResult: String,
    ): String = getHexString(b)

    fun hexStringToByteArray(s: String): ByteArray {
        if (s.length % 2 != 0) {
            throw IllegalArgumentException("Bad input string: $s")
        }

        val len = s.length
        val data = ByteArray(len / 2)
        var i = 0
        while (i < len) {
            data[i / 2] = ((s[i].digitToInt(16) shl 4) + s[i + 1].digitToInt(16)).toByte()
            i += 2
        }
        return data
    }

    fun intToByteArray(value: Int): ByteArray =
        byteArrayOf(
            (value ushr 24).toByte(),
            (value ushr 16).toByte(),
            (value ushr 8).toByte(),
            value.toByte(),
        )

    fun byteArrayToInt(b: ByteArray): Int = byteArrayToInt(b, 0)

    private fun byteArrayToInt(
        b: ByteArray,
        offset: Int,
    ): Int = byteArrayToInt(b, offset, b.size)

    fun byteArrayToInt(
        b: ByteArray,
        offset: Int,
        length: Int,
    ): Int = byteArrayToLong(b, offset, length).toInt()

    fun byteArrayToLong(b: ByteArray): Long = byteArrayToLong(b, 0, b.size)

    fun byteArrayToLong(
        b: ByteArray,
        offset: Int,
        length: Int,
    ): Long {
        if (b.size < offset + length) {
            throw IllegalArgumentException("offset + length must be less than or equal to b.length")
        }

        var value: Long = 0
        for (i in 0 until length) {
            val shift = (length - 1 - i) * 8
            value += (b[i + offset].toInt() and 0x000000FF).toLong() shl shift
        }
        return value
    }

    fun byteArraySlice(
        b: ByteArray,
        offset: Int,
        length: Int,
    ): ByteArray = b.copyOfRange(offset, offset + length)

    fun convertBCDtoInteger(data: Byte): Int = (((data.toInt() and 0xF0) shr 4) * 10) + (data.toInt() and 0x0F)

    fun getBitsFromInteger(
        buffer: Int,
        iStartBit: Int,
        iLength: Int,
    ): Int = (buffer shr iStartBit) and (0xFF shr (8 - iLength))

    /**
     * Reverses a byte array, such that the last byte is first, and the first byte is last.
     *
     * @param buffer     Source buffer to reverse
     * @param iStartByte Start position in the buffer to read from
     * @param iLength    Number of bytes to read
     * @return A new byte array, of length iLength, with the bytes reversed
     */
    fun reverseBuffer(
        buffer: ByteArray,
        iStartByte: Int,
        iLength: Int,
    ): ByteArray {
        val reversed = ByteArray(iLength)
        val iEndByte = iStartByte + iLength
        for (x in 0 until iLength) {
            reversed[x] = buffer[iEndByte - x - 1]
        }
        return reversed
    }

    /**
     * Given an unsigned integer value, calculate the two's complement of the value if it is
     * actually a negative value
     *
     * @param input      Input value to convert
     * @param highestBit The position of the highest bit in the number, 0-indexed.
     * @return A signed integer containing it's converted value.
     */
    fun unsignedToTwoComplement(
        input: Int,
        highestBit: Int,
    ): Int {
        var inp = input
        if (getBitsFromInteger(inp, highestBit, 1) == 1) {
            // inverse all bits
            inp = inp xor ((2 shl highestBit) - 1)
            return -(1 + inp)
        }

        return inp
    }

    // Based on function from mfocGUI by 'Huuf' (http://www.huuf.info/OV/)
    fun getBitsFromBuffer(
        buffer: ByteArray,
        iStartBit: Int,
        iLength: Int,
    ): Int {
        // Note: Assumes big-endian
        val iEndBit = iStartBit + iLength - 1
        val iSByte = iStartBit / 8
        val iSBit = iStartBit % 8
        val iEByte = iEndBit / 8
        val iEBit = iEndBit % 8

        if (iSByte == iEByte) {
            return ((buffer[iEByte].toInt() and 0xFF) shr (7 - iEBit)) and (0xFF shr (8 - iLength))
        } else {
            var uRet = (
                (buffer[iSByte].toInt() and 0xFF and (0xFF shr iSBit)) shl (
                    ((iEByte - iSByte - 1) * 8) +
                        (iEBit + 1)
                )
            )

            for (i in iSByte + 1 until iEByte) {
                uRet = uRet or ((buffer[i].toInt() and 0xFF) shl (((iEByte - i - 1) * 8) + (iEBit + 1)))
            }

            uRet = uRet or ((buffer[iEByte].toInt() and 0xFF) shr (7 - iEBit))

            return uRet
        }
    }

    private val HEX_CHARS =
        charArrayOf(
            '0',
            '1',
            '2',
            '3',
            '4',
            '5',
            '6',
            '7',
            '8',
            '9',
            'a',
            'b',
            'c',
            'd',
            'e',
            'f',
        )
}
